--- a/sql/sql_show.cc	2014-03-25 14:10:46.000000000 +0100
+++ b/sql/sql_show.cc	2014-04-25 11:03:15.000000000 +0200
@@ -2614,6 +2614,9 @@
 // Sends the global table stats back to the client.
 int fill_schema_table_stats(THD* thd, TABLE_LIST* tables, COND* cond)
 {
+  // Disable table statistics.
+  return 0;
+
   TABLE *table= tables->table;
   DBUG_ENTER("fill_schema_table_stats");
   char *table_full_name, *table_schema;
@@ -2658,6 +2661,9 @@
 // Sends the global index stats back to the client.
 int fill_schema_index_stats(THD* thd, TABLE_LIST* tables, COND* cond)
 {
+  // Disable index statistics.
+  return 0;
+
   TABLE *table= tables->table;
   DBUG_ENTER("fill_schema_index_stats");
   char *index_full_name, *table_schema, *table_name;
