From 4bec48f8f022913803267db5587a534f50dc2652 Mon Sep 17 00:00:00 2001
From: Jim Klimov <jimklimov+nut@gmail.com>
Date: Tue, 10 May 2022 16:19:17 +0200
Subject: [PATCH] drivers/dummy-ups.c: use better portable st_mtime [fallout of
 #1383, fix #1415]

---
 drivers/dummy-ups.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/dummy-ups.c b/drivers/dummy-ups.c
index c75cc1e479..6b6399f551 100644
--- a/drivers/dummy-ups.c
+++ b/drivers/dummy-ups.c
@@ -235,7 +235,7 @@ void upsdrv_updateinfo(void)
 					memset(&datafile_stat, 0, sizeof(struct stat));
 					next_update = 1;
 				} else {
-					if (datafile_stat.st_mtim.tv_sec != fs.st_mtim.tv_sec) {
+					if (datafile_stat.st_mtime != fs.st_mtime) {
 						upsdebugx(2, "upsdrv_updateinfo: input file was already read once to the end, but changed later - re-reading");
 						/* updated file => retry ASAP */
 						next_update = 1;
