--- TraceScreen.c.orig	2022-03-13 00:25:30.000000000 +0100
+++ TraceScreen.c	2022-03-13 00:26:03.000000000 +0100
@@ -90,11 +90,21 @@
 
       char buffer[32] = {0};
       xSnprintf(buffer, sizeof(buffer), "%d", this->super.process->pid);
-      execlp("strace", "strace", "-T", "-tt", "-s", "512", "-p", buffer, NULL);
 
-      // Should never reach here, unless execlp fails ...
-      const char* message = "Could not execute 'strace'. Please make sure it is available in your $PATH.";
-      (void)! write(STDERR_FILENO, message, strlen(message));
+      #ifdef __FreeBSD__
+         execlp("truss", "truss",                "-s", "512", "-p", buffer, NULL);
+
+         const char* message = "Could not execute 'truss'. Please make sure it is available in your $PATH.";
+         (void)! write(STDERR_FILENO, message, strlen(message));
+      #endif
+
+      #ifdef __linux__
+         execlp("strace", "strace", "-T", "-tt", "-s", "512", "-p", buffer, NULL);
+
+         // Should never reach here, unless execlp fails ...
+         const char* message = "Could not execute 'strace'. Please make sure it is available in your $PATH.";
+         (void)! write(STDERR_FILENO, message, strlen(message));
+      #endif
 
       exit(127);
    }
