--- SConstruct.orig	2019-05-24 13:09:51.614605000 +0200
+++ SConstruct	2019-05-24 13:10:03.039194000 +0200
@@ -3078,15 +3078,6 @@
         else:
             free_monitoring = "off"
 
-    if not env.TargetOSIs("windows") \
-        and free_monitoring == "on" \
-        and not conf.CheckLibWithHeader(
-        "curl",
-        ["curl/curl.h"], "C",
-        "curl_global_init(0);",
-        autoadd=False):
-        env.ConfError("Could not find <curl/curl.h> and curl lib")
-
     if use_system_version_of_library("pcre"):
         conf.FindSysLibDep("pcre", ["pcre"])
         conf.FindSysLibDep("pcrecpp", ["pcrecpp"])
